pkg/resolver/multi.go

```diff
@@ -0,0 +1,133 @@
+	resolvers resolverMap
+	// ForceDefault will force all names to be resolved by the default
+	// resolution chain, regadless of their TLD.
+	ForceDefault bool
+}
+
+// Option is a function that applies an option to a MultiResolver.
+type Option func(*MultiResolver)
+
+// NewMultiResolver will return a new MultiResolver instance.
+func NewMultiResolver(opts ...Option) *MultiResolver {
+	mr := &MultiResolver{
+		resolvers: make(resolverMap),
+	}
+
+	// Apply all options.
+	for _, o := range opts {
+		o(mr)
+	}
+
+	return mr
+}
+
+// WithForceDefault will force resolution using the default resolver
+// chain.
+func WithForceDefault() Option {
+	return func(mr *MultiResolver) {
+		mr.ForceDefault = true
+	}
+}
+
+// PushResolver will push a new Resolver to the name resolution chain for the
+// given TLD.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will push
+// to the default resolver chain.
+func (mr *MultiResolver) PushResolver(tld string, r Interface) error {
+	if tld != "" && !isTLD(tld) {
+		return ErrInvalidTLD
+	}
+
+	mr.resolvers[tld] = append(mr.resolvers[tld], r)
+	return nil
+}
+
+// PopResolver will pop the last reslover from the name resolution chain for the
+// given TLD.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will pop
+// from the default resolver chain.
+func (mr *MultiResolver) PopResolver(tld string) error {
+	if tld != "" && !isTLD(tld) {
+		return ErrInvalidTLD
+	}
+
+	l := len(mr.resolvers[tld])
+	if l == 0 {
+		return ErrResolverChainEmpty
+	}
+	mr.resolvers[tld] = mr.resolvers[tld][:l-1]
+	return nil
+}
+
+// ChainCount retruns the number of resolvers in a resolver chain for the given
+// tld.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will
+// return the number of resolvers in the default resolver chain.
+func (mr *MultiResolver) ChainCount(tld string) int {
+	return len(mr.resolvers[tld])
+}
+
+// GetChain will return the resolution chain for a given TLD.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will
+// return all resolvers in the default resolver chain.
+func (mr *MultiResolver) GetChain(tld string) []Interface {
+	return mr.resolvers[tld]
+}
+
+// Resolve will attempt to resolve a name to an address.
+// The resolution chain is selected based on the TLD of the name. If the name
+// does not end in a TLD, the default resolution chain is selected.
+// The resolution will be performed iteratively on the resolution chain,
+// returning the result of the first Resolver that succeeds. If all resolvers
+// in the chain return an error, the function will return an ErrResolveFailed.
+func (mr *MultiResolver) Resolve(name string) (Address, error) {
+	tld := ""
+	if !mr.ForceDefault {
+		tld = getTLD(name)
+	}
+	chain := mr.resolvers[tld]
+
+	if len(chain) == 0 {
+		return Address{}, ErrResolverChainEmpty
+	}
+
+	for _, res := range chain {
+		adr, err := res.Resolve(name)
+		if err == nil {
```

Comment:
This error will not be propagated and could be useful after all resolvers fail.

Reply:
Good point. Which error should be propagated (as there is a whole chain of errors) - the last one?

Reply:
The last one would be good enough.